### DISCUSSION TITLE
draw text "on screen"

### DISCUSSION THREAD

**Discussion by farfromrefug** [visgl/deck.gl#5652] [Q&A]
Today i have this working with a `deck.gl`
<img width="369" alt="Screenshot 2021-04-11 at 12 07 42" src="https://user-images.githubusercontent.com/655344/114300073-c5ec6600-9abe-11eb-99ea-f936e1cd2986.png">

Now i would like to achieve something more like this. By like that i mean drawing texts "onscreen" (fixed font size not depending on "distance" to camera). And maybe even draw them at a "z" position depending on the camera view. That way i could draw them let s say at 3000m. But if the camera view dont "see" 3000m i could draw them at a lower z position.
<img width="492" alt="Screenshot 2021-04-11 at 12 09 06" src="https://user-images.githubusercontent.com/655344/114300105-ddc3ea00-9abe-11eb-8c22-c6a96562190f.png">

As anyone any idea on how to achieve that in deck.gl?

**Comment by Pessimistress** [comment]
You can calculate the on-screen positions using [viewport.project](https://deck.gl/docs/api-reference/core/web-mercator-viewport#project) and then render the texts in an OrthographicView on top of the MapView.

See [using multiple views](https://deck.gl/docs/developer-guide/views#using-multiple-views)

**Reply by farfromrefug** [reply]
@Pessimistress nice i see !
I set up a code pen to test it https://codepen.io/farfromrefuge/pen/NWdyBmR
Now i used the `project` inside the `TextLayer` `getPosition`. The issue is that it does not get updated as i move around in the map. I do understand why. What would be the best way for it to dynamically update as i move around?
I return 0,0 right now because i cant seem to find the right "scale" of x,y for the `OrtographicView` 
Seems like 0,0 is the center of the screen.

**Reply by farfromrefug** [reply]
@Pessimistress i am really sorry to bother you but i am not really managing to get this to work. I have update my codepen quite a bit as i understood a lot more things. Still i cant get my "texts" to be "projected" on screen (the main point being to have the same font size whatever the depth position).
I have read somewhere about `modelMatrix` and that it could be changed at render time? Maybe this could help to "project" lat/lon positions to screen, not sure...

**Reply by farfromrefug** [reply]
I actually discovered there was a slack community. Better to ask it there.
Thansk

**Reply by Pessimistress** [reply]
- you need to call `project` from the viewport of the `mainView` (input lng/lat output screen coordinates), not the OrthographicView (input screen coordinates output screen coordinates).
- `getPositions` are only called once unless you invalidate the positions with [updateTriggers](https://deck.gl/docs/api-reference/core/layer#updatetriggers)
- you should listen to Deck's `onViewStateChange` in which you update `currentViewport` and call `setLayers` to update the update triggers. `layerFilter` is called during the render stage, after layer updates are already finished. See https://deck.gl/docs/developer-guide/custom-layers/layer-lifecycle#layer-lifecycle-stages

**Reply by Pessimistress** [reply]
Alternatively you can [subclass](https://deck.gl/docs/developer-guide/custom-layers/subclassed-layers) the [MultiIconLayer](https://github.com/visgl/deck.gl/blob/8.4-release/modules/layers/src/index.js#L38) and overwrite the projected y and z in the vertex shader. This would be the most performant approach but requires some knowledge about writing shaders.

**Reply by farfromrefug** [reply]
@Pessimistress actually yes this is what i thought too and tried. I looked [here](https://github.com/visgl/deck.gl/blob/1049956d723fe7a92c39ea203bc2b5c5fe00cc36/docs/api-reference/core/project.md#glsl-functions) and tried to change the method in the `IconLayer` [here](https://github.com/visgl/deck.gl/blob/8.4-release/modules/layers/src/icon-layer/icon-layer-vertex.glsl.js#L81) But i could not get the method to work with the arguments.
